### ISSUE TITLE
Raise last exception vs always raise RetryError

### ISSUE THREAD

**Issue by bodenr** [jd/tenacity#24]
Today, tenacity will always raise a RetryError if the callable times out in terms of its retries. This behavior is different from retrying that always raised the last raise exception.

While I agree that the behavior taken by default in tenacity is a good choice, porting code from retrying to tenacity becomes error prone with this change in behavior as any callers in the chain must now expect a RetryError.

Yes callers can access the last exception and reraise it with something like:
```py
        try:
            my_retried_fn()
        except tenacity.RetryError as retry_err:
            ex, tr = retry_err.last_attempt.exception_info()
            six.reraise(type(ex), ex, tr)
```

But this will become tedious to replicate across openstack projects where it's "risky" to now raise the RetryError.

What do you guys think about adding some way to toggle on reraising the last exception in tenacity (a backwards compat flag if you will)?

Or if you want to get crazy, perhaps even a pluggable raise.. e.g.
```py
@retry(raise=raise_last_exception()|raise_first_exception()|raise_exception(my_raise_fn)...)
```

**Comment by bodenr** [comment]
Please note that IMHO this is a serious issue and I've marked all the openstack patches that port retrying to tenactiy as a WIP until we can figure out the approach here.

**Comment by jd** [comment]
This is definitely not an issue, and the more I think about it, the less I see a clean solution to this problem. The thing is that if you have a retry rules which states "retry if the result is None for 10 times", then your function _will_ raise `RetryError`, no matter what.

So it means depending on the retrying condition, the error raised might be different. I don't think it's a good design per say, and I think fixing the function using retrying to handle that condition correctly is the good choice (like you did).

However, we could help by having a `reraise()` method so you could wrote:

```python
try:
    my_retried_fn()
except tenacity.RetryError as retry_err:
    retry_err.reraise()
```

Though if the `RetryError` has no exception (because e.g. it retried 10 times because the result was None) we would _anyway_ raise the `RetryError`.

**Comment by bodenr** [comment]
The way OpenStack consumers are using retrying today (and soon tenacity I hope) can be generalized as:
- Retry some function that performs actions(s). In this case the retried function retries if its action(s) raise an exception. If this times out retrying, consumers expect to get the last raised exception.
- Retry some function gated on a condition(s). In this case the retried function retries until some predicate(s) is True. If it times out retrying/tenacity framework will raise the ``RetryError``. In this case consumers explicitly catch the RetryError and take action (ex: https://github.com/openstack/ironic/blob/master/ironic/drivers/modules/ssh.py#L512).

What I'm proposing is some way to maintain this behavior in tenacity.

While I think the ``reraise()`` you proposed above is a great start, I was also hoping to see some way to specify reraise should be used on failure. If we don't have this "switch" then all consumers in the 1st case above (the most common case) will still need to add a new except block to ``reraise()``.

For example https://github.com/openstack/ironic/blob/master/ironic/conductor/task_manager.py#L255 will have to be rewritten as:
```py
        try:
            reserve_node()
        except tenacity.RetryError as retry_err:
            retry_err.reraise()
```

While this is a minor change, it'll be used in almost all our existing consumers that fall into case 1 above. It would be much less impact to consumers if they could just do:

```py
@retry(reraise_by_default=True,...)
def reserve_node(...):
```

to signify they want tenacity to reraise for them; no other changes needed. If for some reason the last exception is a RetryError, then it's just reraised.

**Comment by jd** [comment]
Yeah, I guess we can have both a `reraise` method and make `tenacity` offer to do it via a kwargs. And if there's no exception to reraise, it can still raise a `RetryError` anyhow.

**Comment by bodenr** [comment]
@jd is that something you plan to do, or were you wanting me to kick it off with a PR?

**Comment by jd** [comment]
@bodenr If you have time to do it, please go ahead. :)

**Comment by bodenr** [comment]
@jd thanks.
I think this issue can be closed now right?
What's the chances we can publish a new release of tenacity to pypi so I can work on getting the upper constraints updated in OpenStack for a new version of tenacity that includes PR27?

**Comment by jd** [comment]
Yeah, I thought you closed it in your commit message, but I'll close it now.
I've released 3.1.0 already with that this morning (unless I failed :)

**Comment by bodenr** [comment]
My bad; I didn't even check pypi. Boy you're fast! Thanks!
